pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java

```diff
@@ -73,9 +77,30 @@
 
     protected abstract CompletableFuture<Boolean> existsFromStore(String path);
 
+    // TaskWrapper, wraps the task to be run.
+    // It records the thread on which the task is running. This is to help log the stack trace of a
+    // long running task.
+    static private class TaskWrapper implements Runnable {
+
+        private Runnable runnableTask;
+
+        @Getter
+        private volatile Thread taskThread;
+
+        public TaskWrapper(Runnable task) {
+            runnableTask = task;
+        }
+        @Override
+        public void run() {
+            taskThread = Thread.currentThread();
+            runnableTask.run();
+        }
+    }
+
     protected AbstractMetadataStore() {
         this.executor = Executors
                 .newSingleThreadExecutor(new DefaultThreadFactory("metadata-store"));
+        this.executorWatchDog = Executors.newSingleThreadExecutor(new DefaultThreadFactory("metadata-store-watchdog"));
```

Comment:
This watch dog is worth discussing.
I think you can open a new pr for this.

Reply:
We cannot do per each single request though. We could try to schedule periodic "keep-alive" tasks in these threads and it if they're not completed in X time, kill the broker.

Reply:
@Jason918 separated out the watchdog changes to a new PR, per your suggestion: https://github.com/apache/pulsar/pull/13130

Reply:
@merlimat can you please add your comments in  https://github.com/apache/pulsar/pull/13130. We can discuss it there.